### DISCUSSION TITLE
Refreshing tokens / interceptors

### DISCUSSION THREAD

**Discussion by iway1** [trpc/trpc#2344] [General]
I’m a react native guy, and to manage auth tokens in react native we have to handle everything in JS because there are no cookies.

Always, we have to account for the possibility of our access token expiring and needing to refresh, then retry the request. Ideally, we take it a step further and pause all requests while a refresh is in transit.

In the past I achieved this elegantly with vanilla react query by creating some axios interceptors that handle retrying, pausing and resuming all requests when refreshing.

But now with tRPC we don’t write any of the request functions, and I don’t see any elegant way to pause then resume all requests during a refresh? Is there any way to have an interceptor, or way to pause requests based on some global variable?

the flow I’m looking for is

Client receives unauthorized request -> client pauses any requests -> client refreshes -> client unpauses requests

all i can think of for now is just retrying requests perpetually during  refresh :(

**Comment by KATT** [comment]
I think this could be achieved with a custom link

**Reply by iway1** [reply]
At first glance it looks like it should work great using something like zustand as a global auth store (since it allows accessing and updating state from outside React), I’ll be implementing this tomorrow for our new project and will report back :)

it is satisfying to me that the interceptors can be achieved all within tRPC, while in the past it was necessary for me to build out this layer manually. I think it’s actually just going to be a lot better than how I had been doing it, if it does work out.

thanks KATT!

**Reply by KATT** [reply]
I haven't tried but I really hope that the link architecture should be able to allow you to enable this sort of stuff, the v10 has been rewritten and is a bit more flexible.

See search the source code for dedupeLink for an example. On my phone so a bit hard to link ☺

**Reply by iway1** [reply]
It ended up working great with a custom link in V9! Hadn't use the link architecture before but it was pretty interesting to learn and not too difficult.

Was able to create a pretty cool reusable function that generates a `trpcClient` plus a zustand store where the `trpcClient` will automatically attach headers based on the tokens in the store. 

The resultant API looks like:

```ts
const { trpcClient, useAuthStore } = createAuthClient({
    url: "http://localhost:3005/api/trpc",
    refreshConfig: {
        path: 'auth.generateAccessToken',
        getAccessTokenFromResult: (result) => result.accessToken,
        getParamsForRetry: (token) => ({ token }),
    }
})
```
And then in the components we can just call `login` to login to the store:

```ts
const login = useAuthStore(s => s.login)

const submit = () => {
    loginEmailMutation.mutateAsync({email, password}).then((response)=>{
        login({
            accessToken: response.accessToken,
            refreshToken: response.refreshToken
        })
    })
}
```

Any time there's some unauthorized request while the store has a token, the custom link will try refreshing with the refresh token in the store. The link also pauses all requests during a refresh attempt, and will retry any requests after a successful refresh, and logs the user out in the store if refresh fails or auth in general fails at any point. All this in just one custom link.

Plus the store makes it easy to check if the user is logged in, etc:

```ts
const isLoggedIn = useAuthStore(s => s.isLoggedIn);
```

So yeah, really happy with this! Easy to setup and use in components, and handles everything behind the scenes.

I don't really see myself needing to use links too often just base on the types of apps we've built in the past, but for stuff like this they seem great

**Reply by ru-van-urk** [reply]
Very cool this! Do you have an example of how you created this custom function with the store?

**Reply by iway1** [reply]
@ru-van-urk Sure! I'll just share the code: 

```ts
import { TRPCLink } from "@trpc/client";
import { trpc } from "src/trpc/client";
import { AppRouter } from "../../../api/src/routers";
import { httpLink } from "@trpc/client/links/httpLink";
import { z } from 'zod';
import create from 'zustand';
import { InferMutationInput, InferMutationOutput, TMutation } from "src/trpc/trpc-types";

export const TRPCErrorEnum = z.enum([
  'BAD_REQUEST',
  'UNAUTHORIZED',
  'FORBIDDEN',
  'NOT_FOUND',
  'TIMEOUT',
  'CONFLICT',
  'PRECONDITION_FAILED',
  'PAYLOAD_TOO_LARGE',
  'METHOD_NOT_SUPPORTED',
  'CLIENT_CLOSED_REQUEST',
  'INTERNAL_SERVER_ERROR',
]);

export const TRPCErrorSchema = z.object({
  data: z.object({
    httpStatus: z.number(),
    code: TRPCErrorEnum
  }),
  shape: z.object({
    message: z.string().optional(),
  })
})

interface AuthStore {
  login: (vars: {
    accessToken: string,
    refreshToken: string,
  }) => void,
  logout: () => void,
  accessToken: string | null,
  refreshToken: string | null,
  isLoggedIn: boolean,
}

export function createAuthClient<RetryTokenPath extends TMutation>({
  url,
  refreshConfig
}: {
  url: string,
  refreshConfig: {
    path: RetryTokenPath,
    getAccessTokenFromResult: (result: InferMutationOutput<RetryTokenPath>) => string,
    getParamsForRetry: (refreshToken: string) => InferMutationInput<RetryTokenPath>;
  }
}) {

  // trpc client stuff
  var refreshingRef: { promise: null | Promise<void> } = {
    promise: null
  }

  async function refresh() {
    const params = refreshConfig.getParamsForRetry(useAuthStore.getState().refreshToken!)
    useAuthStore.setState({ accessToken: null })
    //@ts-ignore idk
    const result = await trpcClient.mutation(refreshConfig.path, params);
    const token = refreshConfig.getAccessTokenFromResult(result);
    useAuthStore.setState({
      accessToken: token
    })
  }

  var authLink: TRPCLink<AppRouter> = (runtime) => {
    return async ({ prev, next, op }) => {
      // pause requests while refreshing
      await refreshingRef.promise;
      console.log("\n\nSending request: " + op.path)
      next(op, async (result) => {
        // response from server
        const parse = TRPCErrorSchema.safeParse(result);
        if (!parse.success) {
          // Not an error
          prev(result);
          return;
        }
        if (!(parse.data.data.code == 'UNAUTHORIZED')) {
          prev(result);
          return;
        }
        if (op.path == refreshConfig.path) {
          prev(result);
          return;
        }
        // try to refresh
        const token = useAuthStore.getState().refreshToken;
        if (token === null) prev(result);
        if (refreshingRef.promise !== null) {
          // It's already refreshing, await and retry.
          await refreshingRef.promise;
          next(op, (result) => {
            return prev(result);
          })
        } else {
          // initiate a refresh.
          const p = refresh();
          refreshingRef.promise = p;
          p.then(() => {
            next(op, (result) => {
              return prev(result);
            })
          }).catch(() => {
            prev(result);
            useAuthStore.getState().logout();
          }).finally(() => {
            refreshingRef.promise = null;
          })
        }
      });
    };
  };

  const trpcClient = trpc.createClient({
    url,
    links: [authLink, httpLink({ url })],
    headers: () => {
      var headers: { [key: string]: string } = {};
      const accessToken = useAuthStore.getState().accessToken;
      if (accessToken) {
        console.log("Attaching auth headers.")
        headers['authorization'] = `Bearer ${accessToken}`
      };
      return headers;
    }
  });

  var useAuthStore = create<AuthStore>((set) => {
    return {
      accessToken: null,
      refreshToken: null,
      isLoggedIn: false,
      logout: () => {
        set({
          accessToken: null,
          refreshToken: null,
          isLoggedIn: false,
        });
      },
      login: ({ accessToken, refreshToken }) => {
        set({
          accessToken,
          refreshToken,
          isLoggedIn: true,
        })
      }
    }
  })

  return {
    useAuthStore,
    trpcClient
  }
}
```
The trick here is that we await on the refreshing promise. 

If you just call `login` with the store it'll handle setting all the headers and all that stuff. One known issue with this implementation is that there's a race condition that could cause multiple refreshes to happen when only one was necessary, but it's not the end of the world.
